### ISSUE TITLE
代码中好像有一些问题，会导致最终聚合的客户端与防御方法检测的客户端不一致

### ISSUE THREAD

**Issue by Dxpsk** [JiiahaoXU/MASA#2]
# 代码中好像有一些问题，会导致最终聚合的客户端与防御方法检测的客户端不一致
```python
    def agg_masa(self, agent_updates_dict, flat_global_model):
        local_updates = []
        benign_id = []
        malicious_id = []

        for _id, update in agent_updates_dict.items():
            local_updates.append(update)
            if _id < self.args.num_corrupt:
                malicious_id.append(_id)
            else:
                benign_id.append(_id)

        chosen_clients = malicious_id + benign_id

        stacked_updates = torch.stack(local_updates, dim=0)
        temp_model = copy.deepcopy(self.global_model)
        unlearning_accumulated_loss = []
        pre_aggregation = self.agg_avg(agent_updates_dict)
```
这里的chosen_clients中的客户端顺序和local_updates中的更新客户端顺序不匹配，chosen_clients中的客户端id顺序会将malicious_id放到前面，而正常的id顺序则是随机选取的。
```python
    benign_dict = {}
    for idx in benign_set:
        benign_dict[chosen_clients[idx]] = stacked_updates[idx]

    return self.agg_avg(benign_dict)
```
这里使用chosen_clients[idx]进行索引明显是错误的。同样的，在代码中实现的mul_metric也有这个问题，这导致了mul_metric方法在论文中表现出无法防御任何攻击的情况。鉴于代码是错误的，我很好奇为什么论文中的方法有着良好的防御表现。

**Comment by Dxpsk** [comment]
Sorry, I didn't notice that the IDs were sorted at the beginning of the code.

**Comment by JiiahaoXU** [comment]
> Sorry, I didn't notice that the IDs were sorted at the beginning of the code.

No worry, 实际上这是我们代码的一点小问题（为了更简单的implementation），很抱歉造成了您的误解。

如果您在使用我们的代码时有任何其他问题，或者关于MASA论文的任何问题，欢迎您联系我：）

Regards,
Jiahao

**Comment by JiiahaoXU** [comment]
对于mul_metric方法没有defense效果，我个人的结论是：在其原文中，作者设置了不同的learning rate for benign and malicious clients（应该是在附录中有提到），这样导致the divergence of malicious and benign model updates天然就非常大，从而检测malicious变得容易。

在我们的case里，learning rate for benign and malicious clients始终是保持一致的，增加了检测的难度，从而导致了mul_metric失效。

Regards,
Jiahao
